File: src/main/java/hk/ust/cse/comp4321/proj1/vsm/DocVectorUtils.java
package hk.ust.cse.comp4321.proj1.vsm;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class DocVectorUtils {
    private DocVectorUtils() {}

    public static <T, U, R> Map<String, R> combine(Map<String, T> first, Map<String, U> second, BiFunction<T, U, R> combineFunction, T firstDefault, U secondDefault) {
        return Stream.of(first.keySet(), second.keySet())
                .flatMap(Collection::stream)
                .distinct()
                .collect(Collectors.toMap(x -> x, x -> combineFunction.apply(
                        first.getOrDefault(x, firstDefault),
                        second.getOrDefault(x, secondDefault)
                )));
    }

    public static <T, U, R> Map<String, R> combineIfExists(Map<String, T> first, Map<String, U> second, BiFunction<T, U, R> combineFunction) {
        Set<String> keySet = new HashSet<>(first.keySet());
        keySet.retainAll(second.keySet());
        return keySet.stream()
                .collect(Collectors.toMap(x -> x, x -> combineFunction.apply(
                        first.get(x),
                        second.get(x)
                )));
    }

    public static <T> Map<String, T> combine(Map<String, T> first, Map<String, T> second, BiFunction<T, T, T> combineFunction, T defaultValue) {
        return combine(first, second, combineFunction, defaultValue, defaultValue);
    }

    public static <T, R> Map<String, R> map(Map<String, T> v, Function<T, R> mappingFunction) {
        return v.entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        x -> mappingFunction.apply(x.getValue())
                ));
    }

    public static Map<String, Double> add(Map<String, Double> first, Map<String, Double> second) {
        return combine(first, second, Double::sum, 0.);
    }

    public static Map<String, Double> elementwiseProduct(Map<String, Double> first, Map<String, Double> second) {
        return combineIfExists(first, second, (x, y) -> x * y);
    }

}

File: src/main/kotlin/hk/ust/cse/comp4321/proj1/vsm/DocVectorUtils.kt
package hk.ust.cse.comp4321.proj1.vsm

import java.util.function.BiFunction
import java.util.function.Function

/**
 * The `DocVectorUtils` class is an utility class
 * that contains utility methods for handling document vectors
 * to adopt vector space model in the search engine
 */
object DocVectorUtils {

    fun <T, U, R> combine(first: Map<String, T>, second: Map<String, U>, combineFunction: BiFunction<T, U, R>, firstDefault: T, secondDefault: U): Map<String, R> {
        return first.keys.plus(second.keys)
                .associateWith { combineFunction.apply(first[it] ?: firstDefault, second[it] ?: secondDefault) }
    }

    fun <T> combine(first: Map<String, T>, second: Map<String, T>, combineFunction: BiFunction<T, T, T>, default: T): Map<String, T> {
        return combine(first, second, combineFunction, default, default)
    }

    fun <T, U, R> combineIfExists(first: Map<String, T>, second: Map<String, U>, combineFunction: BiFunction<T, U, R>): Map<String, R> {
        val keySet: MutableSet<String> = HashSet(first.keys)
        keySet.retainAll(second.keys)
        return keySet.associateWith { combineFunction.apply(first[it]!!, second[it]!!) }
    }

    @JvmStatic
    fun <T, R> map(v: Map<String, T>, mappingFunction: Function<T, R>): Map<String, R> {
        return v.mapValues { mappingFunction.apply(it.value) }
    }

    @JvmStatic
    fun add(first: Map<String, Double>, second: Map<String, Double>): Map<String, Double> {
        return combine(first, second, { x, y -> x + y }, 0.0)
    }

    @JvmStatic
    fun elementwiseProduct(first: Map<String, Double>, second: Map<String, Double>): Map<String, Double> {
        return combineIfExists(first, second) { x, y -> x * y }
    }

}
